### DISCUSSION TITLE
Mixing -Parallel, dot sourcing and $using: dont work. Please, explain

### DISCUSSION THREAD

**Discussion by vcudachi** [PowerShell/PowerShell#21348] [Q&A] [answered]
This code is working as intended:
`$v1=10`
`0..2 | ForEach-Object -Parallel {Function F1 {$using:v1};F1}`
But this code is not working:
`$v1=10`
`'Function F1 {$using:v1}' | Set-Content F1.ps1`
`0..2 | ForEach-Object -Parallel {. .\F1.ps1;F1}`
The error is:
**A Using variable cannot be retrieved. A Using variable can be used only with Invoke-Command, Start-Job, or InlineScript in the script workflow. When it is used with Invoke-Command, the Using variable is valid only if the script block is invoked on a remote computer.**

What's the problem here? The documentation says that dot sourcing executes code in the current scope. So this should work.

**Comment by jborean93** [comment, marked as answer]
The variables for $using: must be provided as arguments to the code that is running. PowerShell achieves this by scanning the AST of the scriptblock provided to `-Parallel` and finds any `$using:` statements. In your example the `$using:` statement is inside a separate script which is invoked when the block is run rather than being included at parse time so PowerShell has no way of knowing about it when parsing the original scriptblock to run.

**Reply by vcudachi** [reply]
Thank you! Thus, a corresponding parameter is needed for **_ForEach-Object -Parallel_** so that I can explicitly list the variables that should be available in the scriptblock.
